dictdiffer/__init__.py

```diff
@@ -43,82 +53,89 @@ def diff(first, second, node=None, ignore=None):
        Added *ignore* parameter.
     """
     node = node or []
-    if all(map(lambda x: isinstance(x, string_types), node)):
-        dotted_node = '.'.join(node)
-    else:
-        dotted_node = list(node)
-
-    differ = False
-
-    if isinstance(first, dict) and isinstance(second, dict):
-        # dictionaries are not hashable, we can't use sets
-        def check(key):
-            """Test if key in current node should be ignored."""
-            return ignore is None \
-                or (dotted_node + [key] if isinstance(dotted_node, list)
-                    else '.'.join(node + [str(key)])) not in ignore
-
-        intersection = [k for k in first if k in second and check(k)]
-        addition = [k for k in second if k not in first and check(k)]
-        deletion = [k for k in first if k not in second and check(k)]
+    dotted_node = create_dotted_node(node)
 
-        differ = True
+    def get_extractor(extractors, first, second, node, try_default):
+        extractor = (extractors.get(tuple(node)) or
+                     (extractors.get(type(first))
+                      if type(first) == type(second)
+                      else None))
 
-    elif isinstance(first, list) and isinstance(second, list):
-        len_first = len(first)
-        len_second = len(second)
-
-        intersection = list(range(0, min(len_first, len_second)))
-        addition = list(range(min(len_first, len_second), len_second))
-        deletion = list(reversed(range(min(len_first, len_second), len_first)))
-
-        differ = True
-
-    elif isinstance(first, set) and isinstance(second, set):
+        if extractor:
+            return extractor
+        else:
+            if try_default:
+                for extractor in extractors['default']:
+                    if extractor.is_applicable(first, second):
+                        return extractor
 
-        intersection = {}
-        addition = second - first
-        if len(addition):
-            yield ADD, dotted_node, [(0, addition)]
-        deletion = first - second
-        if len(deletion):
-            yield REMOVE, dotted_node, [(0, deletion)]
+        return None
```

Comment:
This is kind of explicit. Please consider if it wouldn't be better to remove it.